hourly-tips/src/main/java/org/apache/flink/training/exercises/hourlytips/HourlyTipsExercise.java

```diff
@@ -75,19 +84,45 @@ public JobExecutionResult execute() throws Exception {
         // start the data generator
         DataStream<TaxiFare> fares = env.addSource(source);
 
-        // replace this with your solution
-        if (true) {
-            throw new MissingSolutionException();
-        }
+        SingleOutputStreamOperator<Tuple3<Long, Long, Float>> totalDriverTipsPerHourStream = fares
+                .assignTimestampsAndWatermarks(
+                        // taxi fares are in order
+                        WatermarkStrategy.<TaxiFare>forMonotonousTimestamps()
+                                .withTimestampAssigner(
+                                        (fare, t) -> fare.getEventTimeMillis()))
+                .keyBy(fare -> fare.driverId)
+                .window(TumblingEventTimeWindows.of(Time.hours(1)))
+                .process(new CalculateMaxTipPerDriver());
+
+        DataStream<Tuple3<Long, Long, Float>> hourlyMax = totalDriverTipsPerHourStream
+                .keyBy(touple -> touple.f0)
```

Comment:
Your solution works.

The first half is calculating the total amount of tips earned by a driver each hour, which is fine. The result is an un-keyed stream containing the total amount of tips per hour for all drivers. It is sufficient to split this stream into 1 hour windows and find the maximum value.

In the second half, you used another .keyBy method to partition the stream by the tuple's window time. My guess is that you wanted to use the .window method next, which only works on keyed streams. In this case it works, because the key you selected - the window end timestamp - happens to be the same for all the tuples in the stream emitted in the same hour. That means that for each one-hour window, there would only be a single key after using keyBy, making keyBy useless.

KeyBy is an expensive operation, and in this case it is not necessary, because there is no need to do a logical grouping by window end time, this is automatically done by the new windowing.

So, have a look at windowAll methods, which work on non-keyed streams: https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream/operators/windows/#windows
You can continue with your same reduce method on the resulting AllWindowedStream.

Reply:
done, very interesting. thanks for letting me know!